battle.go

```diff
@@ -1,115 +1,212 @@
 package pokemonbattlelib
 
 import (
-	"log"
+	"fmt"
 )
 
 type Battle struct {
-	Weather  int  // one of the 6 in-battle weather conditions
-	ShiftSet bool // shift or set battle style for NPC trainer battles
-	State    BattleState
-	Players  []*Agent
+	Weather      int  // one of the 6 in-battle weather conditions
+	ShiftSet     bool // shift or set battle style for NPC trainer battles
+	State        BattleState
+	Parties      []*Party
+	Agents       []*Agent
+	agentParties map[int]int // Maps agent indexes to party indexes
+	teams        [][]int     // An array of teams, which are arrays of Agents. Used to derive allies and opponents
 }
 
-type BattleState int
+type Party struct {
+	Pokemon []*Pokemon
+	active  []int
+}
 
-type Player int
+func (p *Party) AddPokemon(pkmn ...*Pokemon) {
+	p.Pokemon = append(p.Pokemon, pkmn...)
+}
+
+func (p *Party) SetActive(idx int) {
+	if len(p.active) == 0 {
+		p.active = append(p.active, idx)
+	} else {
+		p.active[0] = idx
+	}
+}
+
+func (p *Party) GetActive() []int {
+	return p.active
+}
+
+// Gives an Agent control over a Party.
+func (b *Battle) LinkAgentParty(agentIdx, partyIdx int) error {
+	if b.agentParties == nil {
+		b.agentParties = make(map[int]int)
+	}
+	// TODO: check if agentIdx, partyIdx are valid indexes
+	// TODO: check if agent is already linked
+	// TODO: check if party is already linked
+	b.agentParties[agentIdx] = partyIdx
+	return nil
+}
+
+// Set which agents have which allies. Not to be confused with `Party`.
+func (b *Battle) SetTeams(t [][]int) error {
+	// TODO: validate
+	b.teams = t
+	return nil
+}
```

Comment:
by naming this `SetTeams()` I imagine you had multi battles in mind (2 Trainers vs. 2 Trainers). do you think this will be able to cleanly support standard double battles (each trainer sends out 2 Pokemon from their party)?

Reply:
It should, because we can allow parties to have multiple active pokemon. I didn't implement that for the sake of keeping this PR as simple as possible.